File: awesometts/service/sapi5js.js
/*
 * AwesomeTTS text-to-speech add-on for Anki
 * Copyright (C) 2010-Present  Anki AwesomeTTS Development Team
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Really simple JScript gateway for talking to the Microsoft Speech API.
 *
 * cscript sapi5js.js voice-list
 * cscript sapi5js.js speech-output <output_path> <rate> <volume> <quality>
 *                                  <flags> <voice_in_hex> <phrase_in_hex>
 */

/*globals WScript*/


var argv = WScript.arguments;

if (typeof argv !== 'object') {
    throw new Error("Unable to read from the arguments list");
}


var argc = argv.count();

if (typeof argc !== 'number' || argc < 1) {
    throw new Error("Expecting the command to execute");
}


var command = argv.item(0);
var options = {};

if (command === 'voice-list') {
    if (argc > 1) {
        throw new Error("Unexpected extra arguments for voice-list");
    }
} else if (command === 'speech-output') {
    if (argc !== 8) {
        throw new Error("Expecting exactly 7 arguments for speech-output");
    }

    var getWavePath = function (path) {
        if (path.length < 5 || !/\.wav$/i.test(path)) {
            throw new Error("Expecting a path ending in .wav");
        }

        return path;
    };

    var getInteger = function (str, lower, upper, what) {
        if (!/^-?\d{1,3}$/.test(str)) {
            throw new Error("Expected an integer for " + what);
        }

        var value = parseInt(str, 10);

        if (value < lower || value > upper) {
            throw new Error("Value for " + what + " out of range");
        }

        return value;
    };

    var getUnicodeFromHex = function (hex, what) {
        if (typeof hex !== 'string' || hex.length < 4 || hex.length % 4 !== 0) {
            throw new Error("Expected quad-chunked hex string for " + what);
        }

        var i;
        var unicode = [];

        for (i = 0; i < hex.length; i += 4) {
            unicode.push(parseInt(hex.substr(i, 4), 16));
        }

        return String.fromCharCode.apply('', unicode);
    };

    // See also sapi5js.py when adjusting any of these
    options.file = getWavePath(argv.item(1));
    options.rate = getInteger(argv.item(2), -10, 10, "rate");
    options.volume = getInteger(argv.item(3), 1, 100, "volume");
    options.quality = getInteger(argv.item(4), 4, 39, "quality");
    options.flags = getInteger(argv.item(5), 0, 16, "flags");
    options.voice = getUnicodeFromHex(argv.item(6), "voice");
    options.phrase = getUnicodeFromHex(argv.item(7), "phrase");
} else {
    throw new Error("Unrecognized command sent");
}


var sapi = WScript.createObject('SAPI.SpVoice');

if (typeof sapi !== 'object') {
    throw new Error("SAPI does not seem to be available");
}


var voices = sapi.getVoices();

if (typeof voices !== 'object') {
    throw new Error("Voice retrieval does not seem to be available");
}

if (typeof voices.count !== 'number' || voices.count < 1) {
    throw new Error("There does not seem to be any voices installed");
}


var i;

if (command === 'voice-list') {
    WScript.echo('__AWESOMETTS_VOICE_LIST__');

    var getHexFromUnicode = function (unicode) {
        if (typeof unicode !== 'string' || !unicode.length) { return ''; }

        var i = 0;
        var chunk;
        var hex = [];

        for (i = 0; i < unicode.length; ++i) {
            chunk = unicode.charCodeAt(i).toString(16);
            switch (chunk.length) {
                case 4:  break;
                case 3:  chunk = '0' + chunk;   break;
                case 2:  chunk = '00' + chunk;  break;
                case 1:  chunk = '000' + chunk; break;
                default: throw new Error("Bad chunk from toString(16) call");
            }

            hex.push(chunk);
        }

        return hex.join('');
    };

    for (i = 0; i < voices.count; ++i) {
        var name;
        try { name = voices.item(i).getAttribute('name'); }
        catch (e) { }

        var language;
        try { language = voices.item(i).getAttribute('language'); }
        catch (e) { }

        if (typeof name == 'string' && name.length > 0) {
          WScript.echo(
              getHexFromUnicode(name) + ' ' +
              (
                  language && getHexFromUnicode(language) ||
                  getHexFromUnicode('unknown')
              )
          );
        }
    }
} else if (command === 'speech-output') {
    var found = false;
    var voice;

    for (i = 0; i < voices.count; ++i) {
        voice = voices.item(i);

        if (voice.getAttribute('name') === options.voice) {
            found = true;
            sapi.voice = voice;
            break;
        }
    }

    if (!found) {
        throw new Error("Could not find the specified voice.");
    }

    var audioOutputStream = WScript.createObject('SAPI.SpFileStream');

    if (typeof audioOutputStream !== 'object') {
        throw new Error("Unable to create an output file");
    }

    audioOutputStream.format.type = options.quality;
    audioOutputStream.open(options.file, 3 /* SSFMCreateForWrite */);

    sapi.audioOutputStream = audioOutputStream;
    sapi.rate = options.rate;
    sapi.volume = options.volume;

    if (options.flags) {
      sapi.speak(options.phrase, options.flags);
    } else {
      sapi.speak(options.phrase);
    }
}

File: awesometts/service/sapi5js.py
# -*- coding: utf-8 -*-

# AwesomeTTS text-to-speech add-on for Anki
# Copyright (C) 2010-Present  Anki AwesomeTTS Development Team
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
Service implementation for SAPI 5 on the Windows platform via JScript

This module functions with the help of a JScript gateway script. See
also the sapi5js.js file in this directory.
"""

import os
import os.path

from .base import Service
from .common import Trait

__all__ = ['SAPI5JS']


LANGUAGE_CODES = {
    '404': 'zh',
    '405': 'cs',
    '406': 'da',
    '407': 'de',
    '408': 'el',
    '409': 'en',
    '40a': 'es',
    '40c': 'fr',
    '410': 'it',
    '411': 'jp',
    '412': 'ko',
    '413': 'nl',
    '415': 'pl',
    '416': 'pt',
    '41d': 'sv',
    '41f': 'tr',
    '436': 'af',
    '439': 'hi',
    '804': 'zh',
}


class SAPI5JS(Service):
    """
    Provides a Service-compliant implementation for SAPI 5 via JScript.
    """

    __slots__ = [
        '_binary',        # path to the cscript binary
        '_voice_list',    # list of installed voices as a list of tuples
    ]

    NAME = "Microsoft Speech API JScript"

    TRAITS = [Trait.TRANSCODING]

    _SCRIPT = os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        'sapi5js.js',
    )

    def __init__(self, *args, **kwargs):
        """
        Attempts to locate the cscript binary and read the list of
        voices from the `cscript.exe sapi5js.js voice-list` output.

        However, if not running on Windows, no environment inspection is
        attempted and an exception is immediately raised.
        """

        if not self.IS_WINDOWS:
            raise EnvironmentError("SAPI 5 is only available on Windows")

        super(SAPI5JS, self).__init__(*args, **kwargs)

        self._binary = next(
            fullpath
            for windows in [
                os.environ.get('SYSTEMROOT', None),
                r'C:\Windows',
                r'C:\WinNT',
            ]
            if windows and os.path.exists(windows)
            for subdirectory in ['syswow64', 'system32', 'system']
            for filename in ['cscript.exe']
            for fullpath in [os.path.join(windows, subdirectory, filename)]
            if os.path.exists(fullpath)
        )

        output = [
            line.strip()
            for line in self.cli_output(
                self._binary,
                self._SCRIPT,
                'voice-list',
            )
        ]

        output = output[output.index('__AWESOMETTS_VOICE_LIST__') + 1:]

        def hex2uni(string):
            """Convert hexadecimal-escaped string back to unicode."""
            return ''.join(chr(int(string[i:i + 4], 16))
                           for i in range(0, len(string), 4))

        def convlang(string):
            """Get language code"""
            string = hex2uni(string).lower().strip()
            return LANGUAGE_CODES.get(string, string)

        self._voice_list = sorted({
            (voice, voice + ' (' + language + ')')
            for (voice, language) in [
                (
                    hex2uni(tokens[0]).strip(),
                    convlang(tokens[1]).strip()
                )
                for tokens
                in [line.split() for line in output]
            ]
            if voice
        }, key=lambda voice: voice[1].lower())

        if not self._voice_list:
            raise EnvironmentError("No voices in `sapi5js.js voice-list`")

    def desc(self):
        """
        Returns a short, static description.
        """

        count = len(self._voice_list)
        return ("SAPI 5.0 via JScript (%d %s)" %
                (count, "voice" if count == 1 else "voices"))

    def options(self):
        """
        Provides access to voice, speed, volume, and quality.
        """

        voice_lookup = dict([
            # normalized with characters w/ diacritics stripped
            (self.normalize(voice[0]), voice[0])
            for voice in self._voice_list
        ] + [
            # normalized with diacritics converted
            (self.normalize(self.util_approx(voice[0])), voice[0])
            for voice in self._voice_list
        ])

        def transform_voice(value):
            """Normalize and attempt to convert to official voice."""

            normalized = self.normalize(value)

            return (
                voice_lookup[normalized] if normalized in voice_lookup
                else value
            )

        return [
            # See also sapi5js.js when adjusting any of these

            dict(
                key='voice',
                label="Voice",
                values=self._voice_list,
                transform=transform_voice,
            ),

            dict(
                key='speed',
                label="Speed",
                values=(-10, 10),
                transform=int,
                default=0,
            ),

            dict(
                key='volume',
                label="Volume",
                values=(1, 100, "%"),
                transform=int,
                default=100,
            ),

            dict(
                key='quality',
                label="Quality",
                values=[
                    (4, "8 kHz, 8-bit, Mono"),
                    (5, "8 kHz, 8-bit, Stereo"),
                    (6, "8 kHz, 16-bit, Mono"),
                    (7, "8 kHz, 16-bit, Stereo"),
                    (8, "11 kHz, 8-bit, Mono"),
                    (9, "11 kHz, 8-bit, Stereo"),
                    (10, "11 kHz, 16-bit, Mono"),
                    (11, "11 kHz, 16-bit, Stereo"),
                    (12, "12 kHz, 8-bit, Mono"),
                    (13, "12 kHz, 8-bit, Stereo"),
                    (14, "12 kHz, 16-bit, Mono"),
                    (15, "12 kHz, 16-bit, Stereo"),
                    (16, "16 kHz, 8-bit, Mono"),
                    (17, "16 kHz, 8-bit, Stereo"),
                    (18, "16 kHz, 16-bit, Mono"),
                    (19, "16 kHz, 16-bit, Stereo"),
                    (20, "22 kHz, 8-bit, Mono"),
                    (21, "22 kHz, 8-bit, Stereo"),
                    (22, "22 kHz, 16-bit, Mono"),
                    (23, "22 kHz, 16-bit, Stereo"),
                    (24, "24 kHz, 8-bit, Mono"),
                    (25, "24 kHz, 8-bit, Stereo"),
                    (26, "24 kHz, 16-bit, Mono"),
                    (27, "24 kHz, 16-bit, Stereo"),
                    (28, "32 kHz, 8-bit, Mono"),
                    (29, "32 kHz, 8-bit, Stereo"),
                    (30, "32 kHz, 16-bit, Mono"),
                    (31, "32 kHz, 16-bit, Stereo"),
                    (32, "44 kHz, 8-bit, Mono"),
                    (33, "44 kHz, 8-bit, Stereo"),
                    (34, "44 kHz, 16-bit, Mono"),
                    (35, "44 kHz, 16-bit, Stereo"),
                    (36, "48 kHz, 8-bit, Mono"),
                    (37, "48 kHz, 8-bit, Stereo"),
                    (38, "48 kHz, 16-bit, Mono"),
                    (39, "48 kHz, 16-bit, Stereo"),
                ],
                transform=int,
                default=39,
            ),

            dict(
                key='xml',
                label="XML",
                values=[
                    (0, "automatic"),
                    (8, "always parse"),
                    (16, "pass through"),
                ],
                transform=int,
                default=0,
            ),
        ]

    def run(self, text, options, path):
        """
        Converts input voice and text into hex strings, writes a
        temporary wave file, and then transcodes to MP3.
        """

        def hexstr(value):
            """Convert given unicode into hexadecimal string."""
            return ''.join(['%04X' % ord(char) for char in value])

        output_wav = self.path_temp('wav')

        try:
            self.cli_call(
                self._binary,
                self._SCRIPT,
                'speech-output',
                output_wav,
                options['speed'],
                options['volume'],
                options['quality'],
                options['xml'],
                hexstr(options['voice']),
                hexstr(text),  # double dash unnecessary due to hex encoding
            )

            self.cli_transcode(
                output_wav,
                path,
                require=dict(
                    size_in=4096,
                ),
            )

        finally:
            self.path_unlink(output_wav)
